File: API/Trick_API/src/main/java/com/trick_manager/Trick_API/entity/SessionGoal.java
package com.trick_manager.Trick_API.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.Data;
import java.time.LocalDateTime;

@Entity
@Table(name = "session_goals")
@Data
public class SessionGoal {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", insertable = false, updatable = false)
    @JsonIgnore
    private User user;

    @Column(name = "user_id")
    private Long userId;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private String type; // "text" or "trick"

    @Column(name = "trick_id")
    private Long trickId;

    @Enumerated(EnumType.STRING)
    @Column(name = "stance")
    private Stance stance = Stance.REGULAR;

    @Column(name = "target_count")
    private Integer targetCount;

    @Column(name = "current_count")
    private Integer currentCount = 0;

    @Column(name = "timer_duration")
    private Long timerDuration; // in seconds

    @Column(name = "remaining_time")
    private Long remainingTime; // in seconds

    @JsonProperty("completed")
    @Column(name = "is_completed")
    private boolean isCompleted = false;

    @Column(name = "is_paused")
    private boolean isPaused = true;

    @Column(name = "created_at")
    private LocalDateTime createdAt = LocalDateTime.now();
}

File: Application/Skaterz/lib/models/session_goal.dart
enum GoalType { text, trick }

class SessionGoal {
  final int? id;
  final String title;
  final GoalType type;
  final int? trickId;
  final int? targetCount;
  int currentCount;
  final Duration? timerDuration;
  Duration? remainingTime;
  bool isCompleted;
  bool isPaused;

  SessionGoal({
    this.id,
    required this.title,
    required this.type,
    this.trickId,
    this.targetCount,
    this.currentCount = 0,
    this.timerDuration,
    this.remainingTime,
    this.isCompleted = false,
    this.isPaused = true,
  });

  factory SessionGoal.fromJson(Map<String, dynamic> json) {
    return SessionGoal(
      id: json['id'],
      title: json['title'],
      type: json['type'] == 'trick' ? GoalType.trick : GoalType.text,
      trickId: json['trickId'],
      targetCount: json['targetCount'],
      currentCount: json['currentCount'] ?? 0,
      timerDuration: json['timerDuration'] != null ? Duration(seconds: json['timerDuration']) : null,
      remainingTime: json['remainingTime'] != null ? Duration(seconds: json['remainingTime']) : null,
      isCompleted: json['completed'] ?? false,
      isPaused: true, // Default to paused when loaded
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'type': type == GoalType.trick ? 'trick' : 'text',
      'trickId': trickId,
      'targetCount': targetCount,
      'currentCount': currentCount,
      'timerDuration': timerDuration?.inSeconds,
      'remainingTime': remainingTime?.inSeconds,
      'completed': isCompleted,
    };
  }
}
